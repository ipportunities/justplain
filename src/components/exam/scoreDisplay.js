import React from 'react'

export const ScoreDisplay = ({ score, maxScore }) => {
  const redColor = {
    r: 229,
    g: 22,
    b: 22
  }
  const orangeColor = {
    r: 244,
    g: 184,
    b: 20
  }
  const greenColor = {
    r: 22,
    g: 225,
    b: 0
  }

  const [r, setR] = React.useState(0)
  const [g, setG] = React.useState(0)
  const [b, setB] = React.useState(0)

  React.useEffect(() => {
    // Calculate the percentage that we need to use for the color
    let percentage = score / maxScore
    // Define that we have a end and start color
    let endColor, startColor

    if (percentage >= 0.5) {
      // If the percentage is above 50% we want to get the color between orange and green.
      startColor = orangeColor
      endColor = greenColor
      // We need the percentage between 50 and 100 percent so we subtract 50 and multiply by 2
      // eg. if percentage is 75% we want the 50% orange and 50% green so we take (75% - 50%) * 2 = 50%
      percentage -= 0.5
      percentage *= 2
    } else {
      startColor = redColor
      endColor = orangeColor
    }

    // DeltaColor is the difference between the start and end color we use this to add to the start color to get the end color
    const deltaColor = {
      r: endColor.r - startColor.r,
      g: endColor.g - startColor.g,
      b: endColor.b - startColor.b,
    }

    // Set RGB to the start color with delta color * the percentage
    setR(startColor.r + (deltaColor.r * percentage))
    setG(startColor.g + (deltaColor.g * percentage))
    setB(startColor.b + (deltaColor.b * percentage))
  }, [score, maxScore])

  return <div style={ { fontWeight: 'bold', color: `rgb(${ r },${ g },${ b })` } }>
    { `Score: ${ score }/${ maxScore } points (${ Math.round(score / maxScore * 100) }%)` }
  </div>

}