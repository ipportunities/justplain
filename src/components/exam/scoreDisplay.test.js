import React from 'react'
import { shallow } from 'enzyme'
import { ScoreDisplay } from './scoreDisplay'

describe('Exam Score Display', () => {
  it('should render correctly with 0%', () => {
    const component = shallow(<ScoreDisplay score={ 0 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 10%', () => {
    const component = shallow(<ScoreDisplay score={ 10 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 20%', () => {
    const component = shallow(<ScoreDisplay score={ 20 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 30%', () => {
    const component = shallow(<ScoreDisplay score={ 30 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 40%', () => {
    const component = shallow(<ScoreDisplay score={ 40 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 50%', () => {
    const component = shallow(<ScoreDisplay score={ 50 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 60%', () => {
    const component = shallow(<ScoreDisplay score={ 60 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 70%', () => {
    const component = shallow(<ScoreDisplay score={ 70 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 80%', () => {
    const component = shallow(<ScoreDisplay score={ 80 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 90%', () => {
    const component = shallow(<ScoreDisplay score={ 90 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly with 100%', () => {
    const component = shallow(<ScoreDisplay score={ 100 } maxScore={ 100 } />)
    expect(component).toMatchSnapshot()
  })
})
