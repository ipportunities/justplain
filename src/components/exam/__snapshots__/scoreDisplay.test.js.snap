// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exam Score Display should render correctly with 0% 1`] = `
<div
  style={
    Object {
      "color": "rgb(229,22,22)",
      "fontWeight": "bold",
    }
  }
>
  Score: 0/100 points (0%)
</div>
`;

exports[`Exam Score Display should render correctly with 10% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 10/100 points (10%)
</div>
`;

exports[`Exam Score Display should render correctly with 20% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 20/100 points (20%)
</div>
`;

exports[`Exam Score Display should render correctly with 30% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 30/100 points (30%)
</div>
`;

exports[`Exam Score Display should render correctly with 40% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 40/100 points (40%)
</div>
`;

exports[`Exam Score Display should render correctly with 50% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 50/100 points (50%)
</div>
`;

exports[`Exam Score Display should render correctly with 60% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 60/100 points (60%)
</div>
`;

exports[`Exam Score Display should render correctly with 70% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 70/100 points (70%)
</div>
`;

exports[`Exam Score Display should render correctly with 80% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 80/100 points (80%)
</div>
`;

exports[`Exam Score Display should render correctly with 90% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 90/100 points (90%)
</div>
`;

exports[`Exam Score Display should render correctly with 100% 1`] = `
<div
  style={
    Object {
      "color": "rgb(0,0,0)",
      "fontWeight": "bold",
    }
  }
>
  Score: 100/100 points (100%)
</div>
`;
